File: 3. 无重复字符的最长子串/Solution.java
class Solution {

    //Solution 1
    //暴力
    public int lengthOfLongestSubstring1(String s) {
        int result = 0;
        for (int i = 0; i < s.length(); i++) {
            Set<Character> set = new HashSet<>();
            int max = 0;
            for (int j = i; j < s.length(); j++) {
                if (set.contains(s.charAt(j))) {
                    break;
                } else {
                    set.add(s.charAt(j));
                    max += 1;
                }
            }
            result = Math.max(result, max);
        }
        return result;
    }

    //Solution 2
    public int lengthOfLongestSubstring2(String s) {
        int result = 0;
        int left = 0;
        int right = 0;
        Set<Character> set = new HashSet<>();
        while (right < s.length()) {
            if (set.contains(s.charAt(right))) {
                set.remove(s.charAt(left));
                left++;
            } else {
                set.add(s.charAt(right));
                right++;
                result = Math.max(result, right - left);
            }
        }
        return result;
    }

}

File: 3. 无重复字符的最长子串/Solution.py
class Solution:

    #Solution 1 
    #暴力
    def lengthOfLongestSubstring1(self, s: str) -> int:
        result = 0
        for i in range(len(s)):
            count = 0
            hashSet = set()
            for j in range(i, len(s)):
                if s[j] not in hashSet:
                    count += 1
                    hashSet.add(s[j])
                else:
                    break
            result = max(count, result)
        return result

    #Solution 2
    #滑动子串
    def lengthOfLongestSubstring2(self, s: str) -> int:
        left = 0
        right = 0
        result = 0
        hashSet = set()
        while right < len(s):
            if s[right] not in hashSet:
                hashSet.add(s[right])
                right += 1
                result = max(result, right - left)
            else:
                hashSet.remove(s[left])
                left += 1

        return result